flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java

```diff
@@ -213,4 +217,11 @@ private CastRuleProvider addRule(CastRule<?, ?> rule) {
                 .findFirst()
                 .orElse(null);
     }
+
+    private LogicalType unwrapDistinct(LogicalType logicalType) {
+        if (logicalType.is(LogicalTypeRoot.DISTINCT_TYPE)) {
```

Comment:
This is just a suggestion, and if we agree it can also be done as another PR: Maybe we can use the `accept()` with a visitor that extracts the underlying type directly, or return the same type (for the simple types) and can be used for other types, like `ARRAY` & `RAW`.

Reply:
Hm, on second thought this must be done before actually resolving the type, so if it's an `ARRAY` then we don't want to extract the underlying type.

Reply:
I think it's fine to don't deeply unwrap it, for example assume you wanna cast using `ARRAY<DISTINCT<INT>>` to `ARRAY<LONG>`. Because in `ArrayToArrayCastRule` the method `CastRuleProvider#exists` and `CastRuleProvider#resolve` are invoked for the array item type, the `DISTINCT<INT>` will be correctly unwrapped to `INT`. Let me add a test just to make sure this works

Reply:
Thx for adding this test!